dyno-recipes/src/main/java/com/netflix/dyno/recipes/lock/DynoLockClient.java

```diff
@@ -0,0 +1,369 @@
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Consumer;
+
+/**
+ * Client for acquiring locks similar to the redlock implementation https://redis.io/topics/distlock
+ */
+public class DynoLockClient {
+
+    private static final Logger logger = LoggerFactory.getLogger(DynoJedisClient.class);
+
+    private final ConnectionPool pool;
+    private final VotingHostsSelector votingHostsSelector;
+    private final ExecutorService service;
+    private final int quorum;
+    private final double CLOCK_DRIFT = 0.01;
+    private TimeUnit timeoutUnit;
+    private long timeout;
+    private final ConcurrentHashMap<String, String> resourceKeyMap = new ConcurrentHashMap<>();
+
+    public DynoLockClient(ConnectionPool pool, VotingHostsSelector votingHostsSelector, long timeout, TimeUnit unit) {
+        this.pool = pool;
+        this.votingHostsSelector = votingHostsSelector;
+        // Threads for locking and unlocking
+        this.service = Executors.newCachedThreadPool();
+        this.quorum = votingHostsSelector.getVotingSize() / 2 + 1;
+        this.timeout = timeout;
+        this.timeoutUnit = unit;
+        // We want to release all locks in case of a graceful shutdown
+        Runtime.getRuntime().addShutdownHook(new Thread(() -> cleanup()));
+    }
+
+    public void setTimeoutUnit(TimeUnit timeoutUnit) {
+        this.timeoutUnit = timeoutUnit;
+    }
+
+    public void setTimeout(long timeout) {
+        this.timeout = timeout;
+    }
+
+    private static String getRandomString() {
+        return UUID.randomUUID().toString();
+    }
+
+    /**
+     * Gets list of resources which are locked by the client
+     * @return
+     */
+    public List<String> getLockedResources() {
+        return new ArrayList<>(resourceKeyMap.keySet());
+    }
+
+    /**
+     * Release the lock (if held) on the resource
+     * @param resource
+     */
+    public void releaseLock(String resource) {
+        if (!checkResourceExists(resource)) {
+            logger.info("No lock held on {}", resource);
+            return;
+        }
+        CountDownLatch latch = new CountDownLatch(votingHostsSelector.getVotingSize());
+        votingHostsSelector.getVotingHosts().getEntireList().stream()
+                .map(host -> new CheckAndRunHost(host, pool, "del", resource, resourceKeyMap.get(resource)))
+                .forEach(ulH -> CompletableFuture.supplyAsync(ulH, service)
+                        .thenAccept(result -> latch.countDown())
+                );
+        boolean latchValue = false;
+        try {
+            latchValue = latch.await(timeout, timeoutUnit);
+        } catch (InterruptedException e) {
+            logger.info("Interrupted while releasing the lock for resource {}", resource);
+        }
+        if (latchValue) {
+            logger.info("Released lock on {}", resource);
+        } else {
```

Comment:
`else` condition unnecessary. The above `catch` can relay the same information.

Reply:
The catch is if the thread gets interrupted. The else is when the latch hits a time out in which case we return immediately.